Bluetooth: Audio: redone the base subgroup struct.

pin-zephyr opened the pull request:
The struct bt_bap_broadcast_assistant_add_src_param and struct bt_bap_broadcast_assistant_mod_src_param use the struct bt_bap_scan_delegator_subgroup to provide subgroup information to the scan delegator. 

The issue is that the struct bt_bap_scan_delegator_subgroup depend on a Kconfig value that only exists for the Scan Delegator (guarded by CONFIG_BT_BAP_SCAN_DELEGATOR).

Therefore, make a general struct bt_bap_base_subgroup which can be used by scan delegator and broadcast assistant. 

Fixes #68338

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -296,7 +287,7 @@ struct bt_bap_scan_delegator_recv_state {
 
 	/** Subgroup specific information */
-	struct bt_bap_scan_delegator_subgroup subgroups[BT_BAP_SCAN_DELEGATOR_MAX_SUBGROUPS];
+	struct bt_bap_base_subgroup subgroups[CONFIG_BT_BAP_BASE_MAX_SUBGROUPS];
```
Ditto: `CONFIG_BT_BAP_BASE_MAX_SUBGROUPS` is too generic (and we have in fact just removed a similar Kconfig).

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -255,15 +246,15 @@ struct bt_bap_unicast_group;
 
 	/** The metadata */
-	uint8_t metadata[BT_BAP_SCAN_DELEGATOR_MAX_METADATA_LEN];
+	uint8_t metadata[CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE];
```
Hmm, this file may actually be included when `CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE` is undefined. 

We may need to guard the `metadata` field with the Kconfig live we've done other places:
```c
struct bt_audio_codec_cap {
	/** Data path ID
	 *
	 * @ref BT_ISO_DATA_PATH_HCI for HCI path, or any other value for
	 * vendor specific ID.
	 */
	uint8_t path_id;
	/** Whether or not the local controller should transcode
	 *
	 * This effectively sets the coding format for the ISO data path to @ref
	 * BT_HCI_CODING_FORMAT_TRANSPARENT if false, else uses the @ref bt_audio_codec_cfg.id.
	 */
	bool ctlr_transcode;
	/** Codec ID */
	uint8_t id;
	/** Codec Company ID */
	uint16_t cid;
	/** Codec Company Vendor ID */
	uint16_t vid;
#if CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE > 0
	/** Codec Specific Capabilities Data count */
	size_t data_len;
	/** Codec Specific Capabilities Data */
	uint8_t data[CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE];
#endif /* CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE > 0 */
#if defined(CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE)
	/** Codec Specific Capabilities Metadata count */
	size_t meta_len;
	/** Codec Specific Capabilities Metadata */
	uint8_t meta[CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE];
#endif /* CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE */
};
```

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -296,7 +287,7 @@ struct bt_bap_scan_delegator_recv_state {
 
 	/** Subgroup specific information */
-	struct bt_bap_scan_delegator_subgroup subgroups[BT_BAP_SCAN_DELEGATOR_MAX_SUBGROUPS];
+	struct bt_bap_bass_subgroup subgroups[CONFIG_BT_BAP_BASS_MAX_SUBGROUPS];
```
Ditto

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -391,7 +382,7 @@ struct bt_bap_scan_delegator_cb {
 	int (*bis_sync_req)(struct bt_conn *conn,
 			    const struct bt_bap_scan_delegator_recv_state *recv_state,
-			    const uint32_t bis_sync_req[BT_BAP_SCAN_DELEGATOR_MAX_SUBGROUPS]);
+			    const uint32_t bis_sync_req[CONFIG_BT_BAP_BASS_MAX_SUBGROUPS]);
```
What happens here if `CONFIG_BT_BAP_BASS_MAX_SUBGROUPS` is undefined?

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -1359,7 +1350,7 @@ int bt_bap_unicast_client_discover(struct bt_conn *conn, enum bt_audio_dir dir);
 
 /** @brief Abstract Broadcast Audio Source Endpoint (BASE) subgroup structure. */
-struct bt_bap_base_subgroup;
+struct bt_bap_bass_subgroup;
```
```suggestion
struct bt_bap_base_subgroup;
```

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -1436,7 +1427,7 @@ int bt_bap_base_get_bis_indexes(const struct bt_bap_base *base, uint32_t *bis_in
  */
 int bt_bap_base_foreach_subgroup(const struct bt_bap_base *base,
-				 bool (*func)(const struct bt_bap_base_subgroup *subgroup,
+				 bool (*func)(const struct bt_bap_bass_subgroup *subgroup,
```
All the `bt_bap_base` parts need to be reverted to `bt_bap_base_subgroup`

Thalley commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -64,4 +64,18 @@ config TARGET_BROADCAST_CHANNEL
 	  Channel Audio Location to sync to. These corresponds to the bt_audio_location,
 	  supporting mono, left and right channels
 
+config BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE
+	int "BASE Subgroup Maximum Metadata Length"
+	default 32
+	range 0 255
+	help
+	  The maximum metadata length support by the BASS server.
```
This already exists in `kconfig.bap`

Thalley commented on samples/bluetooth/broadcast_audio_sink/Kconfig:
```diff
@@ -64,4 +64,18 @@ config TARGET_BROADCAST_CHANNEL
+	help
+	  The maximum metadata length support by the BASS server.
+
+config BT_BAP_BASS_MAX_SUBGROUPS
+	int "BASE Subgroup Maximum Number of Subgroups support"
+	default 1
+	range 0 31
+	help
+	  The maximum number of BIS subgroups supported.
```
This should be defined in `kconfig.bap` instead of in this sample

Thalley commented on samples/bluetooth/broadcast_audio_source/Kconfig:
```diff
@@ -38,4 +38,18 @@ config USE_USB_AUDIO_INPUT
 	select USB_DEVICE_AUDIO
 	select RING_BUFFER
 
+config BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE
+	int "BASE Subgroup Maximum Metadata Length"
+	default 32
+	range 0 255
+	help
+	  The maximum metadata length support by the BASS server.
+
+config BT_BAP_BASS_MAX_SUBGROUPS
+	int "BASE Subgroup Maximum Number of Subgroups support"
+	default 1
+	range 0 31
+	help
+	  The maximum number of BIS subgroups supported.
```
Ditto: These should be defined in `kconfig.bap` instead

Thalley commented on subsys/bluetooth/audio/Kconfig.bap:
```diff
@@ -53,6 +53,13 @@ config BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE
+config BT_BAP_BASS_MAX_SUBGROUPS
+	int "BASE Subgroup Maximum Number of Subgroups support"
+	default 1
+	range 0 31
```
```suggestion
	range 0 25
```

Due limitations in advertising data, the maximum size of a BASE is 253, the maximum size of all subgroups are 249 and the minimum size of a subgroup is 10 octets, so effectively there can be a maximum of 24 subgroups in a BASE
```suggestion
	range 0 24
```

Thalley commented on subsys/bluetooth/audio/Kconfig.bap:
```diff
@@ -53,6 +53,13 @@ config BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE
+	default 1
+	range 0 31
+	help
+	  The maximum number of BIS subgroups supported.
```
```suggestion
	  The maximum number of subgroups supported.
```

pin-zephyr commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -255,15 +246,17 @@ struct bt_bap_unicast_group;
 
 /** Struct to hold subgroup specific information for the receive state */
-struct bt_bap_scan_delegator_subgroup {
+struct bt_bap_base_subgroup {
```
could you elabrate bit more?   are we gonna make 2 structures: struct bass and struct base to be the exactly same ?

Thalley replied:
They won't be the same :) 

The BASE in the advertising data has a structure like this:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/2520413/e568f328-dc3e-4c3b-abba-b43123d4bd7c)


i.e. a structure where there are 1 or more subgroups, with codec id, config data, meta data, and an array of BIS, each with BIS specific config data as well. 

For BASS the structure is much simpler:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/2520413/767bbadd-86bd-449c-8da6-a1ddbbe9c154)

Here we have an array of 1 or more subgroups, but without codec ID, config data, and the list of BIS indexes have been converted to a bitfield. 

`bt_bap_base_subgroup` represents the former, more complex version, and the `bt_bap_bass_subgroup` should represent the latter, simpler version.

pin-zephyr committed:
Bluetooth: Audio: define bt_bap_bass_subgroup struct.

The Broadcast Assistant API should be fully independent of the scan
delegator. Therefore the new struct bt_bap_bass_subgroup has been
defined to avoid a dependency.

Fixes: #68338

Signed-off-by: Ping Wang <pinw@demant.com>

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -367,7 +362,7 @@ struct bt_bap_scan_delegator_cb {
 			       const struct bt_bap_scan_delegator_recv_state *recv_state,
 			       const uint8_t broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE]);
-
+#if defined(CONFIG_BT_BAP_BASS_MAX_SUBGROUPS)
```
Since `CONFIG_BT_BAP_BASS_MAX_SUBGROUPS` cannot be `0`, do we need the guards? 

We have many other Kconfig options that are guarded by `CONFIG_BT_AUDIO` in this file, so I think it's OK to assume that `CONFIG_BT_BAP_BASS_MAX_SUBGROUPS` is always non-0, and we can remove all the guards for this value

Thalley reviewed (changes requested):
LGTM, but I think we can remove some unnecessary guards

henrikbrixandersen commented:
@Thalley Is this bug fix targeting v3.6.0? If so, please add it to the milestone.
